File: remote/src/main/java/com/kelsos/mbrc/dao/Track.java
package com.kelsos.mbrc.dao;

import com.kelsos.mbrc.RemoteDatabase;
import com.raizlabs.android.dbflow.annotation.Column;
import com.raizlabs.android.dbflow.annotation.ForeignKey;
import com.raizlabs.android.dbflow.annotation.ForeignKeyReference;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.structure.BaseModel;

@Table(databaseName = RemoteDatabase.NAME) public class Track extends BaseModel {
  @Column @PrimaryKey(autoincrement = true) long id;
  @Column private String title;
  @Column private int position;
  @Column @ForeignKey(
      references = {
          @ForeignKeyReference(columnName = "genre_id",
              columnType = Long.class,
              foreignColumnName = "id")
      }, saveForeignKeyModel = false)
  private Genre genre;
  @Column @ForeignKey(
      references = {
          @ForeignKeyReference(columnName = "artist_id",
              columnType = Long.class,
              foreignColumnName = "id")
      }, saveForeignKeyModel = false) private Artist artist;
  @Column @ForeignKey(
      references = {
          @ForeignKeyReference(columnName = "album_artist_id",
              columnType = Long.class,
              foreignColumnName = "id")
      }, saveForeignKeyModel = false)
  private Artist albumArtist;
  @Column @ForeignKey(
      references = {
          @ForeignKeyReference(columnName = "album_id",
              columnType = Long.class,
              foreignColumnName = "id")
      }, saveForeignKeyModel = false)
  private Album album;
  @Column private String year;
  @Column private String path;

  public String getPath() {
    return path;
  }

  public Track setPath(String path) {
    this.path = path;
    return this;
  }

  public String getYear() {
    return year;
  }

  public Track setYear(String year) {
    this.year = year;
    return this;
  }

  public Album getAlbum() {
    return album;
  }

  public Track setAlbum(Album album) {
    this.album = album;
    return this;
  }

  public Artist getAlbumArtist() {
    return albumArtist;
  }

  public Track setAlbumArtist(Artist albumArtist) {
    this.albumArtist = albumArtist;
    return this;
  }

  public Artist getArtist() {
    return artist;
  }

  public Track setArtist(Artist artist) {
    this.artist = artist;
    return this;
  }

  public Genre getGenre() {
    return genre;
  }

  public Track setGenre(Genre genre) {
    this.genre = genre;
    return this;
  }

  public int getPosition() {
    return position;
  }

  public Track setPosition(int position) {
    this.position = position;
    return this;
  }

  public String getTitle() {
    return title;
  }

  public Track setTitle(String title) {
    this.title = title;
    return this;
  }

  public long getId() {
    return id;
  }

  public Track setId(long id) {
    this.id = id;
    return this;
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/data/dao/Track.kt
package com.kelsos.mbrc.data.dao

import com.kelsos.mbrc.data.db.RemoteDatabase
import com.raizlabs.android.dbflow.annotation.Column
import com.raizlabs.android.dbflow.annotation.ForeignKey
import com.raizlabs.android.dbflow.annotation.ForeignKeyReference
import com.raizlabs.android.dbflow.annotation.Index
import com.raizlabs.android.dbflow.annotation.IndexGroup
import com.raizlabs.android.dbflow.annotation.PrimaryKey
import com.raizlabs.android.dbflow.annotation.Table
import com.raizlabs.android.dbflow.kotlinextensions.modelAdapter
import com.raizlabs.android.dbflow.structure.Model

@Table(database = RemoteDatabase::class,
    name = "tracks",
    indexGroups = arrayOf(IndexGroup(number = 1, name = "track_title_index")))
data class Track(
    @Column @Index(indexGroups = intArrayOf(1)) var title: String = "",
    @Column var path: String = "",
    @Column var year: String = "",
    @Column var position: Int = 0,
    @Column var disc: Int = 0,
    @Column
    @ForeignKey(references = arrayOf(ForeignKeyReference(columnName = "genre_id",
        columnType = Long::class,
        referencedFieldIsPrivate = true,
        referencedGetterName = "getId",
        referencedSetterName = "setId",
        foreignKeyColumnName = "id")), saveForeignKeyModel = false)
    var genre: Genre? = null,
    @Column
    @ForeignKey(references = arrayOf(ForeignKeyReference(columnName = "artist_id",
        columnType = Long::class,
        referencedFieldIsPrivate = true,
        referencedGetterName = "getId",
        referencedSetterName = "setId",
        foreignKeyColumnName = "id")), saveForeignKeyModel = false)
    var artist: Artist? = null,
    @Column
    @ForeignKey(references = arrayOf(ForeignKeyReference(columnName = "album_artist_id",
        columnType = Long::class,
        referencedFieldIsPrivate = true,
        referencedGetterName = "getId",
        referencedSetterName = "setId",
        foreignKeyColumnName = "id")), saveForeignKeyModel = false)
    var albumArtist: Artist? = null,
    @Column
    @ForeignKey(references = arrayOf(ForeignKeyReference(columnName = "album_id",
        columnType = Long::class,
        referencedFieldIsPrivate = true,
        referencedGetterName = "getId",
        referencedSetterName = "setId",
        foreignKeyColumnName = "id")), saveForeignKeyModel = false)
    var album: Album? = null,

    @Column(name = "date_added") var dateAdded: Long = 0,
    @Column(name = "date_updated") var dateUpdated: Long = 0,
    @Column(name = "date_deleted") var dateDeleted: Long = 0,
    @Column @PrimaryKey(autoincrement = true) var id: Long = 0) : Model {

  override fun load() = modelAdapter<Track>().load(this)

  override fun insert(): Long = modelAdapter<Track>().insert(this)

  override fun save() = modelAdapter<Track>().save(this)

  override fun update() = modelAdapter<Track>().update(this)

  override fun exists(): Boolean = modelAdapter<Track>().exists(this)

  override fun delete() = modelAdapter<Track>().delete(this)
}
